doc/developer/design/20240625_source_versioning__table_from_sources.md

```diff
@@ -0,0 +1,412 @@
+MySQL sources do not allow querying the source's primary relation and instead
+multiplex into _subsources_, an entirely separate concept. In addition to these
+core source types, we also support the concept of _push sources_ (e.g. the
+Materialize Fivetran connector), which write data to tables.
+
+Our existing _source_ models also make it difficult to deal with upstream schema changes,
+such that users need to drop and recreate a _source_ or _subsource_ and any dependent objects
+to begin ingesting data with a new schema. This can cause long-periods of downtime if the
+new source needs to reingest the entire upstream history. This can also cause them to
+lose their _reclocking_ information which makes it impossible to implement a correct
+end-to-end _source_ -> _sink_ pipeline.
+
+Since subsources are created automagically as the result of a `CREATE SOURCE ..` statement
+it is difficult to manage them with tools like dbt, which require all objects to be
+created and dropped using explicit statements per object.
+
+This change aims to simplify the user experience and create a more unified
+model for managing data ingested from an upstream external system, while providing
+more flexibility in handling upstream schema changes and managing sources in tools like dbt.
+
+## Success Criteria
+
+-   The mental overhead & learning time for users to ingest data from an upstream system
+    is decreased.
+-   All source types use a consistent interface to retrieve progress information and
+    ingested data. The primary _source_ object represents the same concept across
+    all source types.
+-   The same upstream table/topic can be ingested into multiple collections in Materialize,
+    each with a potentially different projection of the upstream schema. This enables
+    a user to do blue-green schema-change workflows without unavailability.
+-   _sources_ can maintain their progress collection while the set of collections being
+    ingested are modified, allowing an end-to-end _source_ -> _sink_ schema-change workflow
+    that does not emit redundant data (reusable reclocking).
+-   The object representing each upstream table/topic in materialize is created and dropped
+    explicitly, such that they can be managed as models directly in dbt.
+
+## Out of Scope
+
+<!--
+What does a solution to this problem not need to address in order to be
+successful?
+
+It's important to be clear about what parts of a problem we won't be solving
+and why. This leads to crisper designs, and it aids in focusing the reviewer.
+-->
+
+-   Accomplishing 'subsource demuxing' using the concept of _subsources_ in-use today.
+    The goals of https://github.com/MaterializeInc/materialize/issues/24843 will be
+    accomplished during the process of replacing _subsources_ with _tables_ (see
+    solution proposal below).
+
+-   Dealing with Webhook sources. These are fundamentally different than the other
```

Comment:
We'll move this to within scope and get help from Adapter with porting webhook sources into the new model. Added a placeholder task to #28653.